Client/Frontend/Widgets/GIFImage.swift

```diff
@@ -0,0 +1,179 @@
+  }
+
+  init(asset: String, animationRepeatCount: Int = 1) {
+    super.init(frame: .zero)
+    self.assetName = asset
+    self.animationRepeatCount = animationRepeatCount
+    initView()
+  }
+  
+  override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+    super.traitCollectionDidChange(previousTraitCollection)
+    loadAsset()
+  }
+  
+  private func loadAsset() {
+    let frames = Self.getFramesFrom(asset: assetName)
+    firstFrame = frames.first
+    lastFrame = frames.last
+    framesCount = frames.count
+    imageView.image = frames.first
+    imageView.animationImages = frames
+    imageView.animationRepeatCount = animationRepeatCount
+  }
+
+  override func layoutSubviews() {
+    super.layoutSubviews()
+    imageView.frame = bounds
+    self.addSubview(imageView)
+  }
+  
+  func startAnimation(start: Bool) {
+    if start {
+      imageView.image = lastFrame
+      imageView.startAnimating()
+    } else {
+      imageView.image = firstFrame
+      imageView.stopAnimating()
+    }
+  }
+  
+  func animationDidStop(_ anim: CAAnimation, finished: Bool) {
+    print(finished)
+  }
+
+  private func initView() {
+    imageView.contentMode = .scaleAspectFit
+    loadAsset()
+  }
+  
+  /// Get `gif` image frames (an array of images) from an asset by its name
+  private class func getFramesFrom(asset: String) -> [UIImage] {
+    guard let asset = getAsset(name: asset) else { return [] }
+    return getFrames(from: asset.data)
+  }
+  
+  /// Get `gif` image asset by its name. Will append `-dark` to the name when dark mode is enabled
+  private class func getAsset(name: String) -> NSDataAsset? {
+    switch UITraitCollection.current.userInterfaceStyle {
+    case .light, .unspecified:
+      return NSDataAsset(name: name, bundle: .current)
+    case .dark:
+      return NSDataAsset(name: [name, "dark"].joined(separator: "-"), bundle: .current)
+    @unknown default:
+      return NSDataAsset(name: name, bundle: .current)
+    }
+  }
+  
+  /// Get `gif` image frames (an array of images) from data
+  class func getFrames(from data: Data) -> [UIImage] {
+    guard let source = CGImageSourceCreateWithData(data as CFData, nil) else {
+      return []
+    }
+    
+    let count = CGImageSourceGetCount(source)
+    let delays = (0..<count).map {
+      // store in ms and truncate to compute GCD more easily
+      Int(delayForImage(at: $0, source: source) * 1000)
+    }
+    
+    let gcd = delays.reduce(0, gcd)
+    
+    return (0..<count).flatMap { index -> [UIImage] in
+      guard let cgImage = CGImageSourceCreateImageAtIndex(source, index, nil) else {
+        return []
+      }
+      let frame = UIImage(cgImage: cgImage)
+      let frameCount = delays[index] / gcd
+      
+      return (0..<frameCount).map { index2 -> UIImage in
+        return frame
+      }
+    }
+  }
+  
+  private class func gcd(_ a: Int, _ b: Int) -> Int {
+    let absB = abs(b)
+    let r = abs(a) % absB
```

Comment:
nit: Can we use one word variables instead of single letters ?

Reply:
sure. I stole this from the internet and didn't do a good job cleaning it up